File: sortVertices.js
function getAngle(v1, v2){
    var dotproduct = v1.x*v2.x + v1.y*v2.y + v1.z*v2.z;
    var mag_v1 = Math.sqrt(v1.x*v1.x + v1.y*v1.y + v1.z*v1.z);
    var mag_v2 = Math.sqrt(v2.x*v2.x + v2.y*v2.y + v2.z*v2.z);
    var cosTheta = dotproduct/(mag_v1*mag_v2);
    return Math.acos(cosTheta);
}

function getNearest(v, vertices){
    var min = 100;
    var minindex = 0;
    for(i=0; i<vertices.length; i++){
        if(vertices[i].x == null)
            continue;
        var angle = getAngle(v, vertices[i]);
        if(angle < min){
            min = angle;
            minindex = i;
        }
    }
    return minindex;
}

function centerize(vertices){
    var mean = {}
    var x = 0;
    var y = 0;
    var z = 0;
    vertices.forEach(vertice => {
        x+=vertice.x;
        y+=vertice.y;
        z+=vertice.z;
    });
    mean['x'] = x/vertices.length;
    mean['y'] = y/vertices.length;
    mean['z'] = z/vertices.length;

    vertices.forEach(vertice => {
        vertice['x'] -= mean.x;
        vertice['y'] -= mean.y;
        vertice['z'] -= mean.z;
    });

}

function sortVertices(vertices){
    var vertices_copy = JSON.parse(JSON.stringify(vertices));
    centerize(vertices_copy);
    var starting_index = 0
    var current = JSON.parse(JSON.stringify(vertices_copy[starting_index]));
    var selected = [starting_index];
    vertices_copy[starting_index]['x'] = null
    for(var i=1; i<vertices_copy.length; i++){
        var nearest = getNearest(current, vertices_copy);
        current = JSON.parse(JSON.stringify(vertices_copy[nearest]));
        vertices_copy[nearest]['x'] = null;
        selected.push(nearest);
    }
    return selected;
}

File: sortVertices.py
import numpy as np
import time

def getNearest(v1, v2, discard):
    v2 = v2.transpose()
    angle =  np.arccos(np.dot(v1, v2) / (np.linalg.norm(v1) * np.linalg.norm(v2)))
    min = 100
    minindex = 0
    for i in range(0, len(angle)):
        if i not in discard and angle[i] < min:
            min = angle[i]
            minindex = i
    return minindex


if __name__ == '__main__':
    v1 = np.array([0, 5, 0])
    v2 = np.array([5, 0, 0])
    v3 = np.array([-5, 0, 0])
    v4 = np.array([0, -5, 0])
    v5 = np.array([3, 3, 0])
    vertices = np.array([v1, v2, v3, v4, v5])
    middle = vertices.mean(axis=0)
    centered_vertices = vertices - middle

    current = centered_vertices[0]
    selected_indices = list()
    selected_indices.append(0)
    for i in range(1, len(centered_vertices)):
        nearest = getNearest(current, centered_vertices, selected_indices)
        selected_indices.append(nearest)
        current = centered_vertices[nearest]

    print(selected_indices)
